### PR TITLE
Allows TC to be ejected from uplinks

### PR DESCRIPTION
**Pull request by Zonespace27**
[Feature] [Balance] [QoL]
## About the PR 
This PR adds the ability to eject "Pure Telecrystals"* from any uplink. These Pure Telecrystals* can be inserted into any unlocked uplink, allowing for traitors, nukeops, and revs to share their TC, for cooperation.

In addition, it also adds a "Trick Pure Telecrystal"* which costs 1 TC, and is visually the same as the regular Pure Telecrystals*, but explodes with the force of a PDA bomb upon anyone trying to insert it into an uplink.

*The Pure Telecrystal's name will be the same name as that round's syndicoins.

## Why's this needed? 
I've felt like this kind of feature has been missing for a while, and I felt that it'd be worth adding. While I do admit, I haven't thought through every possible balance implication of this feature, I still feel like this is a feature worth adding.

## Changelog

```changelog
(u)Zonespace
(*)You may now eject telecrystals from uplinks, to be put in any other unlocked uplink.
(*)There is also an explosive version of these telecrystals, for when you want to be a little less cooperative.
```

### CODE REVIEW

**Comment by DimWhat** [comment]
I am VERY against traitors getting more badguy bucks, it is a recipe for disaster.

**Comment by alexeric123** [comment]
I like this personally, especially how it could pave the way in the future to items that requires two traitors to cooperate to buy.

**Comment by ZeWaka** [comment]
> I am VERY against traitors getting more badguy bucks, it is a recipe for disaster.

literally not possible unless someone gives their bucks to them - which could be done anyways by just giving them their PDA. this is a qol change.

good change

File: code/obj/item/device/pda2/pda2.dm [hunk 1]
```diff
@@ -613,6 +613,21 @@
...
+			var/crystal_amount = C.amount
+			src.uplink.uses = src.uplink.uses + crystal_amount
+			if (crystal_amount == 1)
+				boutput(user, "You insert [crystal_amount] telecrystal into the [src].")
+			else
+				boutput(user, "You insert [crystal_amount] telecrystals into the [src].")
+			qdel(C)
+		else if (!src.uplink.active && !isnull(src.uplink.lock_code))
+			return
+		//return ..()
```
Comment (ZeWaka) [inline code/obj/item/device/pda2/pda2.dm:629]: 👁️ 👅 👁️

File: code/obj/item/device/pda2/pda2.dm [hunk 2]
```diff
@@ -613,6 +613,21 @@
...
+		if (src.uplink.active)
+			var/crystal_amount = C.amount
+			src.uplink.uses = src.uplink.uses + crystal_amount
+			if (crystal_amount == 1)
+				boutput(user, "You insert [crystal_amount] telecrystal into the [src].")
+			else
+				boutput(user, "You insert [crystal_amount] telecrystals into the [src].")
+			qdel(C)
+		else if (!src.uplink.active && !isnull(src.uplink.lock_code))
+			return
```
Comment (ZeWaka) [inline code/obj/item/device/pda2/pda2.dm:628]: useless!!

File: code/obj/item/explosives.dm [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+	icon = 'icons/obj/materials.dmi'
+	icon_state = "telecrystal_ref"
+	pickup(mob/user)
+		boutput(user, "<span class='alert'>The [src] explodes!</span>")
+		var/turf/T = get_turf(src.loc)
+		if(T)
+			T.hotspot_expose(700,125)
+
+			explosion(src, T, -1, -1, 2, 3) //about equal to a PDA bomb
+		src.set_loc(user.loc)
```
Comment (ZeWaka) [inline code/obj/item/explosives.dm:13]:
```suggestion
		if(T)
			T.hotspot_expose(700,125)
			explosion(src, T, -1, -1, 2, 3) //about equal to a PDA bomb

		src.set_loc(user.loc)
```
weird spacing

File: code/obj/item/material.dm [hunk 4]
```diff
@@ -729,6 +741,11 @@
 	default_material = "telecrystal"
 	icon_state = "martian-bar"
 
+/obj/item/material_piece/telecrystal_ref
+	desc = "A cut block of Refined Telecrystal."
+	default_material = "telecrystal_ref"
+	icon_state = "martian-bar"
```
Comment (ZeWaka) [inline code/obj/item/material.dm:747]: 👀

File: code/obj/item/uplinks.dm [hunk 5]
```diff
@@ -198,6 +198,22 @@ Note: Add new traitor items to syndicate_buylist.dm, not here.
...
+				var/crystal_amount = W.amount
+				uses = uses + crystal_amount
+				if (crystal_amount == 1)
+					boutput(user, "You insert [crystal_amount] telecrystal into the [src].")
+				else
+					boutput(user, "You insert [crystal_amount] telecrystals into the [src].")
+				qdel(W)
+			else if (src.locked && !isnull(src.lock_code))
+				return
+		return
```
Comment (ZeWaka) [inline code/obj/item/uplinks.dm:215]: uselessssss

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by Merlin1230** [comment]
I always like cooperating with other traitors, and this makes it much better to do so.

**Comment by ThatFiveGuys** [comment]
Havent checked code but if its the case, I feel like itd be bad for explosive telecrystals to be able to be gotten from surplus's, because new players would see refined telecrystals from there surplus, pick it up, and die. Also it may be interesting to allow surplus's to flat out give you more refined telecrystals (Something like 3 refined telecrystals at the cost of 3 TC for the surplus). idk just tossing that idea out.

**Comment by BatElite** [comment]
Concern of mine: Is having the evil dollars be actual physical telecrystals going to lead to people thinking you can maybe get more of them again? Even if that's not an issue in itself, regular telecrystal can already be refined into blocks through regular methods (the triangle sprot is the raw "ore"), so you may need a different adjective for clarity.

File: code/obj/item/material.dm (hunk 4 above)
Comment (Sovexe) [reply code/obj/item/material.dm:747]: shouldn't be a material at all imo. Just an item.

**Comment by flappybatpal** [comment]
I think this raises too many "what if" scenarios which aren't good.

Metagaming friends, robust antags more easily getting extra TC from antags they kill combine and buy a strong item, valid hunters combining TC etc....

**Comment by ZeWaka** [comment]
> robust antags more easily getting extra TC from antags they kill combine and buy a strong item, valid hunters combining TC etc....

and this here is why you always lock your PDA or use all of your TC

File: code/obj/item/device/pda2/pda2.dm [hunk 6]
```diff
@@ -613,6 +613,18 @@
 				src.insert_id_card(ID, user)
 				boutput(user, "<span class='notice'>You insert [ID] into [src].</span>")
 
+	else if (istype(C, /obj/item/raw_material/telecrystal_ref))
+		if (!C || !C.material)
+			return
+		if (src.uplink.active)
```
Comment (BatElite) [inline code/obj/item/device/pda2/pda2.dm:619]: This would runtime on PDAs without an uplink.

File: code/obj/item/material.dm (hunk 4 above)
Comment (Zonespace27) [reply code/obj/item/material.dm:747]: them being a material allows for easy stacking, having slightly better than telecrystal matsci properties, and lets you make a traitor cry as they watch all their TC get put in a reclaimer

File: code/obj/item/uplinks.dm [hunk 7]
```diff
@@ -36,7 +37,7 @@ Note: Add new traitor items to syndicate_buylist.dm, not here.
 	New()
 		..()
 		SPAWN_DBG(1 SECOND)
-			if (src && istype(src) && (!src.items_general.len && !src.items_job.len && !length(src.items_objective)))
```
Comment (ZeWaka) [inline code/obj/item/uplinks.dm:39]: stop using `.len`

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/obj/item/material.dm (hunk 4 above)
Comment (Sovexe) [reply code/obj/item/material.dm:747]: More concerned mat duping making them unlimited

File: code/obj/item/uplinks.dm (hunk 7 above)
Comment (Zonespace27) [reply code/obj/item/uplinks.dm:39]: removed every instance of .len from uplinks.dm

**Comment by MatheusIsNotAvailable** [comment]
I think the trick telecrystal should explode when inserted in a pda instead of on pick up, would be bad if someone bought it and then couldn't position it right because it is on the ground now, or add a check to see if the person picking it up is not the one that bought it.

**Comment by penny64** [comment]
I think the ejected item should have the same name as the crime tokens in your PDA, for one because the name is funny, and two it avoids the old confusion about TC

**Comment by Azrun** [comment]
I think sharing items is interesting, but I feel like this shifts the balance of some items because it feels like certain combinations are intended to leave you with 1-2 that can't be spent on a more powerful item.  The restriction breeds ingenuity and creativity feels min/max and meta prone.  

I just hope I'm completely wrong because this is a cool idea.

**Comment by PrimeNumb** [comment]
> I think sharing items is interesting, but I feel like this shifts the balance of some items because it feels like certain combinations are intended to leave you with 1-2 that can't be spent on a more powerful item. The restriction breeds ingenuity and creativity while this feels min/max and meta prone.
> 
> I just hope I'm completely wrong because this is a cool idea.

As ZeWaka already pointed out, this doesn't actually change anything balance-wise. You can already give your traitor PDA to someone else.

**Comment by flappybatpal** [comment]
This does have some implications because it's not uncommon for people to spend some TC, partial TC now becomes more valuable and makes those 7+ cost combos that undergo tons of scrutiny much more likely.

I don't really see positive implications. I prefer traitor cooperation at "arms length" where they can work together but nothing directly enables it.

**Comment by PrimeNumb** [comment]
It's true that there's an edge case where if you have leftover TC and find another PDA, then yes, you could circumvent the cost restrictions imposed on the high value traitor items (i.e. items balanced around not being able to be bought together with cloaking device) by pooling the TC into one PDA.

File: code/obj/item/explosives.dm [hunk 8]
```diff
@@ -1,13 +1,13 @@
 /obj/item/explosive_telecrystal
 	name = "pure telecrystal"
-	desc = "A pure telecrystal, useful for creating small, precise warps in space."
+	desc = "A pure Telecrystal, useful for creating small, precise warps in space."
 	icon = 'icons/obj/materials.dmi'
 	icon_state = "telecrystal_pure"
-	pickup(mob/user)
+	/*pickup(mob/user)
```
Comment (ZeWaka) [inline code/obj/item/explosives.dm:6]: 👀

Comment (Zonespace27) [reply code/obj/item/explosives.dm:6]: Fuck, forgot to remove that, sorry!

**Review by Tarmunora — CHANGES_REQUESTED** [review]
Tying this into materials feels unnecessary. This would work just as well having them be generic McGuffins.

**Comment by Zonespace27** [comment]
> Tying this into materials feels unnecessary. This would work just as well having them be generic McGuffins.
@Tarmunora 
It makes it easier to stack them (and process them being in a stack) without having to repeat any unnecessary code, the rest of the material bits I could care less about.

**Comment by Tarmunora** [comment]
Item stacking behaviour is defined at the `obj/item` level iirc

**Review by Tarmunora — CHANGES_REQUESTED** [review]
Need handling to prevent splitting from stack after putting crystals in an uplink etc

**Review by Tarmunora — CHANGES_REQUESTED** [review]
should probably block them from surplus crates as well

**Comment by Zonespace27** [comment]
> should probably block them from surplus crates as well

i felt it'd be interesting to occasionally get a traitorcoin or two out of surpluses, but if you feel otherwise I'll go ahead and change it.

**Review by Tarmunora — APPROVED** [review]

**Pull request merged**